yt/utilities/answer_testing/framework.py

```diff
@@ -63,6 +81,332 @@
+        hex_digests['pixelized_projection_values'] = ppv_hd 
+        return hex_digests
+
+    def generic_array_test(self, func, args=None, kwargs=None):
+        if args is None:
+            args = []
+        if kwargs is None:
+            kwargs = {}
+        return func(*args, **kwargs)
+
+    def sph_answer(self, ds, ds_str_repr, ds_nparticles, field, weight, ds_obj, axis):
+        # Make sure we're dealing with the right dataset
+        assert str(ds) == ds_str_repr
+        # Set up keys of test names
+        hex_digests = {} 
+        dd = ds.all_data()
+        assert dd["particle_position"].shape == (ds_nparticles, 3)
+        tot = sum(dd[ptype, "particle_position"].shape[0]
+                  for ptype in ds.particle_types if ptype != "all")
+        # Check
+        assert tot == ds_nparticles
+        dobj = utils.create_obj(ds, ds_obj)
+        s1 = dobj["ones"].sum()
+        s2 = sum(mask.sum() for block, mask in dobj.blocks)
+        assert s1 == s2
+        if field[0] in ds.particle_types:
+            particle_type = True
+        else:
+            particle_type = False
+        if particle_type is False:
+            ppv_hd = self.pixelized_projection_values_test(ds, axis, field, weight, ds_obj)
+            hex_digests['pixelized_projection_values'] = ppv_hd
+        fv_hd = self.field_values_test(ds, field, ds_obj, particle_type=particle_type)
+        hex_digests['field_values'] = fv_hd
+        return hex_digests
+
+    def yt_field_test(self, ds, field, geometric):
+        if geometric:
+            obj = ds.all_data()
+        else:
+            obj = ds.data
+        return np.array([obj[field].size, obj[field].mean()])
+
+    def plot_window_attribute_test(self, ds, plot_field, plot_axis, attr_name,
+        attr_args, plot_type='SlicePlot', callback_id='', callback_runners=[]):
+        plot = utils.create_plot(ds, plot_type, plot_field, plot_axis, {})
+        for r in callback_runners:
+            r(plot_field, plot)
+        attr = getattr(plot, attr_name)
+        attr(*attr_args[0], **attr_args[1])
+        tmpfd, tmpname = tempfile.mkstemp(suffix='.png')
+        os.close(tmpfd)
+        plot.save(name=tmpname)
+        image = mpimg.imread(tmpname)
+        os.remove(tmpname)
+        return image
+
+    def phase_plot_attribute_test(self, ds_fn, x_field, y_field, z_field,
+                 attr_name, attr_args, plot_type='PhasePlot',
+                 plot_kwargs={}):
+        data_source = ds_fn.all_data()
+        plot = utils.create_plot2(data_source, x_field, y_field,
+                                z_field, plot_type, plot_kwargs)
+        attr = getattr(plot, attr_name)
+        attr(*attr_args[0], **attr_args[1])
+        tmpfd, tmpname = tempfile.mkstemp(suffix='.png')
+        os.close(tmpfd)
+        plot.save(name=tmpname)
+        image = mpimg.imread(tmpname)
+        os.remove(tmpname)
+        return image
```

Comment:
For the image tests was the ultimate decision not to save the hashes of the images? That's fine, just want to double-check.

Reply:
The image data hashes get saved.